aten/src/ATen/native/mps/operations/Quantized.mm

```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+  using simdgroup_type8x8 = simdgroup_bfloat8x8;
+  using type4 = bfloat4;
+};
+#endif
+
+template<typename T>
+float2 get_scale_zero(constant T * scalesAndZeros, uint2 index) {
+    return float2(1.0, 0.0);
+}
+
+template<typename T>
+float2 get_scale_zero_q8(constant T * scalesAndZeros, uint2 index) {
+    T scale = scalesAndZeros[index[0]];
+    return float2(scale, 0.0);
+}
+
+#define BLOCK_SIZE_M 32 // each block takes 32 rows in matrix A
+#define BLOCK_SIZE_N 64 // each block takes 64 rows in matrix B
+#define BLOCK_SIZE_K 32
+#define THREAD_MAT_M 2 // in data loading stage, each thread load 2 simdgroup matrices from matrix A
+#define THREAD_MAT_N 4 // in data loading stage, each thread load 4 simdgroup matrices from matrix B
+#define THREAD_PER_ROW_A 4 // 4 thread for each row in matrix A to load numbers
+#define THREAD_PER_ROW_B 2 // 2 thread for each row in matrix B to load numbers
+#define SG_MAT_SIZE 64 // simdgroup matrix is of shape 8x8
+#define SG_MAT_ROW 8
+
+// T: input type, W: weight type
+template<typename T, typename W, float2 (*get_scale_zero_func)(constant T *, uint2)>
+kernel void kernel_mul_mm(
+    constant T                 * A              [[buffer(0)]],
+    constant char              * B              [[buffer(1)]],
+    constant T                 * scalesAndZeros [[buffer(2)]],
+    device T                   * outputData     [[buffer(3)]],
+    constant uint3             & sizes          [[buffer(4)]],
+    threadgroup char           * shared_memory  [[threadgroup(0)]], // threadgroup buffer at index 0
+    uint3                        tgpig          [[threadgroup_position_in_grid]], // 3d coordinates
+    uint                         tiitg          [[thread_index_in_threadgroup]], // 128 per threadgroup
+    uint                         sgitg          [[simdgroup_index_in_threadgroup]]) {
+
+    using T4 = typename BlockType<T>::type4;
+    using Tsimd8x8 = typename BlockType<T>::simdgroup_type8x8;
+    // sizes: x = M, y = K, z = N
+    // pytorch: M x K @ N x K -> M x N
+    // ggml: K x N @ K x M -> N x M
+    uint32_t M = sizes.x; // M
+    uint32_t K = sizes.y; // K
+    uint32_t N = sizes.z; // N
+    uint32_t nb00 = sizeof(W);
+    uint32_t nb01 = nb00 * K;
+    uint32_t nb10 = sizeof(T);
+    uint32_t nb11 = nb10 * K;
+
+    // 8192 for sa, 4096 for sb
+    threadgroup T      * sa = (threadgroup T      *)(shared_memory);
+    threadgroup bfloat * sb = (threadgroup bfloat *)(shared_memory + 8192);
+
+    const uint r0 = tgpig.y;
+    const uint r1 = tgpig.x;
+
+    // if this block is of 64x32 shape or smaller
+    short n_rows_A = (M - r1 * BLOCK_SIZE_M < BLOCK_SIZE_M) ? (M - r1 * BLOCK_SIZE_M) : BLOCK_SIZE_M;
+    short n_rows_B = (N - r0 * BLOCK_SIZE_N < BLOCK_SIZE_N) ? (N - r0 * BLOCK_SIZE_N) : BLOCK_SIZE_N;
+
+    // a thread shouldn't load data outside of the matrix
+    short thread_row_A = ((short)tiitg/THREAD_PER_ROW_A) < n_rows_A
+        ? ((short)tiitg/THREAD_PER_ROW_A)
+        : n_rows_A - 1;
```

Comment:
can this be simplified to `((short)tiith / THREAD_PER_ROW_A) <= n_rows_A - 1 ?: n_rows_A - 1`, or perhaps an expression with min()?

Reply:
[Elvis operator](https://en.wikipedia.org/wiki/Elvis_operator) is not part of C/C++ standard, nor do I see it mentioned in Metal spec (though it works)